pcl_ros/tools/pcd_to_pointcloud.cpp

```diff
@@ -43,126 +43,89 @@
 
  **/
 
-// ROS core
-#include <ros/ros.h>
-#include <pcl/io/io.h>
-#include <pcl/io/pcd_io.h>
-#include <pcl/point_types.h>
-#include <pcl_conversions/pcl_conversions.h>
-
 // STL
 #include <chrono>
 #include <string>
 #include <thread>
 
-#include "pcl_ros/publisher.hpp"
+// PCL
+#include <pcl/io/io.h>
+#include <pcl/io/pcd_io.h>
+#include <pcl/point_types.h>
+
+// ROS core
+#include <rclcpp/rclcpp.hpp>
+#include "rclcpp_components/register_node_macro.hpp"
+#include <sensor_msgs/msg/point_cloud2.hpp>
+#include <pcl_conversions/pcl_conversions.h>
 
-class PCDGenerator
+namespace pcl_ros
+{
+class PCDGenerator : public rclcpp::Node
 {
 protected:
   std::string tf_frame_;
-  ros::NodeHandle nh_;
-  ros::NodeHandle private_nh_;
 
 public:
   // ROS messages
-  sensor_msgs::PointCloud2 cloud_;
+  sensor_msgs::msg::PointCloud2 cloud_;
 
   std::string file_name_, cloud_topic_;
-  double wait_;
+  size_t period_ms_;
 
-  pcl_ros::Publisher<sensor_msgs::PointCloud2> pub_;
+  std::shared_ptr<rclcpp::Publisher<sensor_msgs::msg::PointCloud2>> pub_;
+  rclcpp::TimerBase::SharedPtr timer_;
 
   ////////////////////////////////////////////////////////////////////////////////
-  PCDGenerator()
-  : tf_frame_("/base_link"), private_nh_("~")
+  explicit PCDGenerator(const rclcpp::NodeOptions & options)
+  : rclcpp::Node("pcd_generator", options), tf_frame_("/base_link")
   {
     // Maximum number of outgoing messages to be queued for delivery to subscribers = 1
 
     cloud_topic_ = "cloud_pcd";
-    pub_.advertise(nh_, cloud_topic_.c_str(), 1);
-    private_nh_.param("frame_id", tf_frame_, std::string("/base_link"));
-    ROS_INFO(
-      "Publishing data on topic %s with frame_id %s.", nh_.resolveName(
-        cloud_topic_).c_str(), tf_frame_.c_str());
-  }
+    tf_frame_ = this->declare_parameter("spin_period", tf_frame_);
+    period_ms_ = this->declare_parameter("publishing_period_ms", 3000);
+    file_name_ = this->declare_parameter<std::string>("file_name");
 
-  ////////////////////////////////////////////////////////////////////////////////
-  // Start
-  int
-  start()
-  {
     if (file_name_ == "" || pcl::io::loadPCDFile(file_name_, cloud_) == -1) {
-      return -1;
+      RCLCPP_ERROR(this->get_logger(), "failed to open PCD file");
+      throw std::runtime_error{"could not open pcd file"};
     }
     cloud_.header.frame_id = tf_frame_;
-    return 0;
+    int nr_points = cloud_.width * cloud_.height;
+
+    auto fields_list = pcl::getFieldsList(cloud_);
+    auto resolved_cloud_topic =
+      this->get_node_topics_interface()->resolve_topic_name(cloud_topic_);
+    RCLCPP_DEBUG(
```

Comment:
Why not make this INFO and then delete the INFO below that is a partial subset of what this says?